service/history/execution/mutable_state_builder.go

```diff
@@ -2379,18 +2346,44 @@ func (e *mutableStateBuilder) logDuplicatedActivityEvents(events []*types.Histor
 				tag.WorkflowRunID(e.GetExecutionInfo().RunID),
 				tag.WorkflowScheduleID(scheduledEventID),
 				tag.WorkflowEventType(event.GetEventType().String()),
-				tag.Dynamic("duplication-source", duplicationSource),
+				tag.Dynamic("duplication-source", source),
 			)
 
 			e.metricsClient.IncCounter(metrics.HistoryFlushBufferedEventsScope, metrics.DuplicateActivityTaskEventCounter)
-		} else {
-			activityTaskUniqueEvents[uniqueEventParams] = struct{}{}
+			return false
 		}
+		activityTaskUniqueEvents[uniqueEventParams] = struct{}{}
+		return true
 	}
 
+	var headEvents []*types.HistoryEvent
+	var tailEvents []*types.HistoryEvent
+
+	// Sometimes we see buffered events are out of order when read back from database.  This is mostly not an issue
+	// except in the Activity case where ActivityStarted and ActivityCompleted gets out of order.  The following code
+	// is added to reorder buffered events to guarantee all activity completion events will always be processed at the end.
 	for _, event := range events {
-		checkActivityTaskEventUniqueness(event)
+		// We sometimes see duplicate events
+		if unique := checkActivityTaskEventUniqueness(event); !unique {
+			continue
+		}
+		switch event.GetEventType() {
+		case types.EventTypeActivityTaskCompleted,
+			types.EventTypeActivityTaskFailed,
+			types.EventTypeActivityTaskCanceled,
+			types.EventTypeActivityTaskTimedOut:
+			tailEvents = append(tailEvents, event)
+		case types.EventTypeChildWorkflowExecutionCompleted,
+			types.EventTypeChildWorkflowExecutionFailed,
+			types.EventTypeChildWorkflowExecutionCanceled,
+			types.EventTypeChildWorkflowExecutionTimedOut,
+			types.EventTypeChildWorkflowExecutionTerminated:
+			tailEvents = append(tailEvents, event)
+		default:
+			headEvents = append(headEvents, event)
```

Comment:
Sorry for the dumb questions, but can you walk me through the implications of new event X being added to the Head here? What is the happy case / bad case?

Reply:
Head events will just be the events in their original order, minus any that are moved to the tail.

A happy case scenario with events [started, completed] will get moved to head:[started], tail:[completed], and then merged back to [started, completed].

The case we're addressing is out of order events: [completed, started], which  gets moved to head:[started], tail:[completed], and then merged back to [started, completed]. But of course with multiple sets of activities it gets more complicated.